### ISSUE TITLE
Snes9x 1.56ish

### ISSUE THREAD

**Issue by bearoso** [snes9xgit/snes9x#297]
With all the changes lately, it's getting time for a new release. We need to come up with a list of TODOs to accomplish before it's ready.

* @OV2, you mentioned getting the GLSL shader support into Windows. I might drop the Cg support from Gtk+ since I did it so haphazardly. I've been looking at the new slang format, but that can wait until the next release.

* I'd like to get support for reading the monitor's true refresh rate into the Windows port so we don't have to rely on the user to set the input rate or even know what it is. For reference, I added something like this to Retroarch a while back using QueryDisplayConfig. In Gtk+ I've got it set by default to just poll the monitor and not need input rate adjustment.

* I *did* add a GUI config to the Gtk+ port to adjust the SuperFX overclock. I don't know if you want to expose that to Windows users.

**Comment by jjfmz42015** [comment]
Add menu option to enable\disable BlockInvalidVRAMAccess maybe? If done add note recommended on but may fix issues when disabled?

**Comment by bearoso** [comment]
Obviously fixing the games would be the ideal solution, but I’d prefer just adding the games to a hack list before making that flag visible.

**Comment by jjfmz42015** [comment]
ok so add kick off to a hack list, and another thought would be to move fps display to black border when rendering in 4:3 in a wide screen resolution.

Just throwing ideas out here not demanding. Im already super happy with Snes9x and appreciate having this emulator.

**Comment by OV2** [comment]
I'll try to find some time this or next weekend to give the glsl shaders a shot.

For the refresh rate: you want to query this and then set up the input rate accordingly? I'd still want to expose the input rate, since it might be running in auto mode without vsync. So maybe an auto-adjust button that queries and sets the slider?
Also shouldn't the dynamic rate control be ok in most cases?

I'm not sure about the superfx overclock setting. I guess I don't mind putting a gui setting somewhere.

**Comment by Oggom** [comment]
Are there any plans to get in sync with the libretro repository again? It appears the last attempt was made here https://github.com/snes9xgit/snes9x/pull/60, but it never went anywhere.

**Comment by OV2** [comment]
I ported some of the changes for 1.55. I'll see if there have been any since then that make sense for us - I'm not sure about fast savestates and hard audio disable, for example.

**Comment by jjfmz42015** [comment]
Is there an option to hide the mouse pointer if using a mouse as a gun device? Ill check this when i get home. That is always handy in emulators when using a modern lcd gun. I own a top gun ii but never tried it in snes9x yet. I have spent most of my time in snes 9x playing mario world and all stars and yoshis island and final fight and arkanoid and r type latley and others but only with my snes usb pads. For the longest time i was a zsnes user many years ago then jumped to snes9x then snes gt then snes9x then bsnes now back to snes9x. I find the current dev build of snes9x for windows to be my favorite

**Comment by bearoso** [comment]
>For the refresh rate: you want to query this and then set up the input rate accordingly? I'd still want to expose the input rate, since it might be running in auto mode without vsync. So maybe an auto-adjust button that queries and sets the slider?
Also shouldn't the dynamic rate control be ok in most cases?

What I'm doing now with the Gtk+ port is using a checkbox to automatically set things by default and sets the slider accordingly. Unchecking it lets the user change it at will. I had an auto-adjust button there instead, but I wasn't sure if the user would know to hit it when first configuring the emulator.

There was a thread on the libretro forums about the dynamic rate control being audible, and the samples provided showed I was susceptible to it. Also, it won't very well with large variations, so making sure it's centered properly helps.

>I'm not sure about the superfx overclock setting. I guess I don't mind putting a gui setting somewhere.

I noticed when debugging Doom that our default timings were too slow, but it also seemed correct for Star Fox. Previously I had noticed Yoshi's Island 2 running too slowly. Because some games may run too fast and others may run too slowly with our SuperFX code, I figured I'd let the user configure it. It's up to you.

**Comment by bearoso** [comment]
@OV2, let me know if you start working on something. Since I've gotten my feet wet working with Windows, I can jump in and do some stuff like the input rate thing.

> and maybe add static display or some sort of logo or animation when rom is not loaded like punes, zsnes, kega fusion if you get what i mean.

You should see the funny thing I did with the Gtk+ port--a static logo with a bunch of patterns to test filters on.
![screenshot from 2018-05-19 18-41-16](https://user-images.githubusercontent.com/16159164/40274105-79e60cc8-5b94-11e8-883c-bfb5932de3b3.png)

**Comment by OV2** [comment]
I'll start with the shaders tomorrow - if you want to give the input rate a shot be my guest :)

**Comment by Felipefpl** [comment]
Hi, i didnt get 1.55, only 1.54 because... no will to do it :P - and i saw something strange:

Why 2 changelogs, .rft and .txt? I mean, only one is enough (a simple .txt file like it exists in other emulators would do) but the .txt changelog is a mess to read while the .rft is not, excluding one and prettyfing the other would be a simple solution for this. ;)

**Comment by jjfmz42015** [comment]
Geez i thought i was picky

**Comment by OV2** [comment]
I've made a first commit that includes the glsl shaders, although it's not really working at the moment. One problem is the POT texture used in the win port for the main output. Since regular and hi-res can have different scaling I'd have to change the texture size on the fly, I'll have to think about that some more. Besides that the shaders don't seem to affect the image at all, even though all gl functions succeed. I'll get back on it later.

**Comment by bearoso** [comment]
It looks like you're still calling DrawArrays after you call the glsl render function, so maybe it's overwriting the image? I had the glsl render process use the framebuffer as output for the last pass to avoid that extra viewport-sized copy at the end, which differs from how you had Cg doing it and is why the viewport is passed to the function. 

I put a glLogErrors function in shader_helpers that you can call somewhere during shader init and it'll log all the normally hidden OpenGL stuff to stderr. Could be helpful.

**Comment by OV2** [comment]
Thanks, that was the problem. I'm already using glIntercept to get ogl errors, but the logging might be useful in the future.

One issue with the current implementation: you can't change the viewport depending on the output size of the last pass. For example when using the plain hq2x shader and you want it centered in the window instead of stretched. Any ideas? A viewport callback maybe?

I haven't switched the CG code yet since you mentioned you might drop it.

**Comment by bearoso** [comment]
I didn't quite understand what you were doing with the viewport in the Cg code. A viewport callback or port function might be necessary. I've already got something in the form of S9xApplyAspect in the GTK+ port, and you've got CalculateDisplayRect. I'd prefer something a little more portable than either, though. I think a port function something like
`S9xCalculateDisplayViewport(int width, int height, int *outx, int *outy, int *outwidth, int *outheight)`
with pointers instead of references would be the most versatile.

**Comment by bearoso** [comment]
I should point out that the Retroarch people added a bunch of stuff to the Cg shader spec since you added support. There's stuff like passprev, aliases, mipmapping, and a bunch of other things used by the common shader repository. I started to add those, but realized instead of trying to fill out our Cg support, we'd probably be better off putting work towards moving to the future-proof slang format.

**Comment by Pat86** [comment]
Since fixing the PAL version of Secret of Evermore (+ Super Punch Out?) - as a major title for the machine - got fixed and should be flawlessly playable now, why not bump the version to 1.60? This would fetch more attention as well.

**Comment by bearoso** [comment]
I simply cannot skip 56. https://youtube.com/watch?v=Qs8kDiOwPBA

**Comment by qwertymodo** [comment]
Not only are there fixes for those major titles, this release also fixes Speedy Gonzales, one of byuu's go-to examples for comparing less accurate emulators against higan.  Between that and the complete cheats overhaul to a completely new data format, I kind of agree with the idea of 1.6...

**Comment by bearoso** [comment]
Eh, Speedy Gonzales is an interesting edge case, but an inaccurate emulator can definitely get away with it. There obviously wasn’t any real accuracy improvement other than for a very specific situation. This just means byuu needs a different example.

**Comment by qwertymodo** [comment]
Spoken like a true engineer with no eye for marketing.  In addition to the other points, we also have that snazzy new icon...

**Comment by bearoso** [comment]
>One issue with the current implementation: you can't change the viewport depending on the output size of the last pass. For example when using the plain hq2x shader and you want it centered in the window instead of stretched. Any ideas? A viewport callback maybe?

I added a viewport callback to the GLSL render function. You still have to pass in what you think is the final viewport because some shaders use that in interrim stages. The viewport callback parameters contain both the rendered image size and the viewport parameters it has used for rendering those interrim passes. You set the output parameters to your desired glViewport params.

**Comment by bearoso** [comment]
Also, in case it's confusing: make sure you check the parameters passed into render. I switched the order of viewport x, y and width, height.

**Comment by OV2** [comment]
I've changed the OGL code to use NPOT textures if available and used the viewport callback, looks good now. I'm still working on the parameter dialog, but that shouldn't be too hard.

Regarding the new CG stuff: I saw that a while ago, but since the spec was never updated and it seemed somewhat fluid I never added anything. I'd like to keep it around so that there is at least something for D3D, but I'm not sure if it is worth the trouble updating since many of the new changes did not even work in RA D3D-mode.

**Comment by OV2** [comment]
I've added a first version of a glsl parameter settings dialog that seems to work fine.

Have you decided if you want to keep cg in? I wouldn't mind switching to a central implementation.

**Comment by jjfmz42015** [comment]
call 1.56 1.56ish instead like this subject.

**Comment by jjfmz42015** [comment]
new logo? with purple buttons to match new icon? 

https://drive.google.com/file/d/1TUqdmd8HjSffGi6fTMGCernxCfaGpc85/view?usp=sharing

**Comment by jjfmz42015** [comment]
maybe a crazy idea but what if you added the ability to allow a user to switch to a liberto core for a specific erom if they wanted?

example use snes 9x and if a specific title worked better on let say bsnes core user can choose to load and switch to that core.

No something I really want but just a suggestion if you guys are looking for ideas to add to a next major build.

kinda how bizhawk lets you switch between snes9x and bsnes

again just a thought as a cool feature. dont know if anyone would really care for it.

**Comment by jjfmz42015** [comment]
also the custom open rom menu. make it bigger, make it possible for a user to define a static rom directory that always remains static.

good for installing for non savvy users who might "lose" their rom directory.

**Comment by bearoso** [comment]
Have you ever heard of a phrase called "feature creep?"

**Comment by ImSpecial** [comment]
So with all these recent shader changes, what does this exactly mean? ELI5 me on it if you can. I hope we don't lose the current shaders that are already in Snes9x, I happen to like a lot of them and there is basically no setup required.

**Comment by OV2** [comment]
It means we additionally support .glsl and .glslp shader files. Nothing has been removed.

**Comment by ImSpecial** [comment]
Cool, in addition is always good, I was a little worried that the old shaders were going in favor of newer ones.

**Comment by bearoso** [comment]
They're the same shaders in a different format. In fact, they're a little more compatible because the Cg format was secretly updated several times and we never noticed.

**Comment by bearoso** [comment]
At this point what is there left to do for 1.56?

**Comment by bearoso** [comment]
I noticed my IRQ changes broke Battle Blaze's title screen. I'll have to look into that.

**Comment by OV2** [comment]
I'll add the fx overclock setting, and I guess we should see if someone is willing to compile the osx port again.

And you'll have to decide how the release should be named ;)

**Comment by bearoso** [comment]
Battle Blaze is entering an IRQ handler, reading 4211 to clear the IRQ line, but then sets another IRQ timer that occurs during that handler and pulls the IRQ line low again so that it triggers instantly upon leaving. We can allow it to pull the IRQ line while the handler is active, but that breaks Marko's Magic Football. Shrug.

**Comment by OV2** [comment]
Mortal Kombat 2 also seems broken: the character models randomly break up for a short time. Might be a similar issue.

**Comment by bearoso** [comment]
I've pushed a fix for Battle Blaze. See if this fixes Mortal Kombat 2.

**Comment by OV2** [comment]
Unfortunately id didn't change anything for mk2.

IIRC from the higan sources the irq line is held for 4 cycles when it triggers, and blocked for a while after certain events (e.g. writing 4200), but I can't remember for how long.

**Comment by bearoso** [comment]
Did Mortal Kombat 2 have problems before? From what I found by byuu a loooong time ago, MK2 is very "picky." I'd have to analyze it.

Marko isn't a problem with IRQ timing, AFAICT. What SimpleTease/sluffy say in issue #231 is true. It reads from 4211 with a data bank of 7F during the interrupt handler, which is WRAM. I'll have to look at where that 7F is coming from.

**Comment by bearoso** [comment]
Yeah, mortal kombat is *exactly* like the problem byuu had back in 2007: http://forums.nesdev.com/viewtopic.php?t=2932

**Comment by OV2** [comment]
MK2 worked previously because of the delay in CheckInterrupts
```
static inline void S9xCheckInterrupts (void)
{
	bool8	thisIRQ = PPU.HTimerEnabled | PPU.VTimerEnabled;

	if (CPU.IRQLine & thisIRQ)
		CPU.IRQTransition = TRUE;
```

If I add something similar to the new way MK2 works again, but then battle blaze exhibits its problem.

**Comment by bearoso** [comment]
Nah, I fixed it. It's exactly like byuu's problem. WAI needed to be more granular.

**Comment by bearoso** [comment]
According to byuu:
> For instance, Marko's Magic Football will clear an IRQ and still expect it to trigger right after. This works because the two-stage pipeline performs the work-cycle IRQ test one cycle behind the bus-cycle next-opcode fetch. The same does not hold true for clearing the IRQ with REP, because that instruction clears the flags more than one cycle prior to the opcode termination. Attempting to simulate this pipeline is quite the challenge, you have to turn one serial thread into two parallel processes.

**Comment by bearoso** [comment]
> If I add something similar to the new way MK2 works again, but then battle blaze exhibits its problem.

After fixing my mistake, I had to bump up the time after WAI to 14 cycles to fix MK2, which means I probably need to implement the transition time. It didn't exactly work before--it was still just factored into the trigger time, and IRQTransition was just a flag--but I'll look into it over the next couple days. I'm not sure if we need the hold delay, as I doubt many games are reading 4211 for anything but resetting the IRQ line, especially immediately.

**Comment by bearoso** [comment]
After the OpenGL thing, what else do we have left for 1.56?

**Comment by OV2** [comment]
I can't think of any pressing issues. Maybe checking with byuu about the .bml file usage?

The osx port will probably be broken because of the cheat changes.

**Comment by bearoso** [comment]
byuu said it was fine. I don't know about the OS X port. We obviously can't just wait until someone appears who can update it. I could make some blind code changes, but I imagine some errors would creep in.

**Comment by bearoso** [comment]
Should we go ahead and tag the release, then?

**Comment by bearoso** [comment]
Ok, so I tested all the IRQ games I know about and they all still work.

**Comment by bearoso** [comment]
Shall we try tagging again? I'll go ahead and do it tomorrow if nothing major comes in.

**Comment by OV2** [comment]
Sure. I think I can find some time tonight to build and upload.

**Comment by bearoso** [comment]
Release tagged. I also added the changes to a github release.

**Comment by qwertymodo** [comment]
MSU-1 stereo bug should be in the changelog.

**Comment by bearoso** [comment]
I updated the release page and put it into changes.txt.

**Comment by bearoso** [comment]
There. No more revisionist history from this point on. :-)

**Comment by OV2** [comment]
I've mirrored the source and built win versions: http://s9x-w32.de/dl/

**Comment by bearoso** [comment]
Who wants to make a post to the forums, then?

**Comment by OV2** [comment]
I've made a post in the forum and messaged Ryan to make it sticky.

**Comment by bearoso** [comment]
I might do a hotfix 1.56.1. A couple fellows mentioned the automatic input rate was breaking sound, and I suspect it's because they have non-60hz displays and I didn't think to check that.

**Comment by bearoso** [comment]
I'm able to build the Windows binaries in that case, to avoid you the hassle.

**Comment by OV2** [comment]
Can you build with directdraw enabled? I'll check if I can maybe include the required stuff so that this also works with the automated builds...

What I've never really thought about: does the automatic input rate make sense if vsync is disabled and it uses the automatic framerate limit?

**Comment by bearoso** [comment]
I can do DirectDraw. I don't think Appveyor will have the old SDK in their image, though.

> What I've never really thought about: does the automatic input rate make sense if vsync is disabled and it uses the automatic framerate limit?

I don't really think automatic input rate makes sense, which is why I just set 0 frameskip on GTK+. We're using 16.666ms for NTSC games? That's kind of wrong. If you're using the true SNES input rate (32040) it'll underflow. byuu just has sound sync and video sync. Maybe we should follow suit.

Personally, I think the automatic input rate is a good default because with vsync enabled sound is smooth and video is smooth. If you think it should be disabled by default, definitely go ahead and do so.

**Comment by OV2** [comment]
By default it's using PCBase / 59.948743718592964824120603015098f

The automatic framerate limit is also used for speed up / slow down - I don't really want to get rid of that feature.

**Comment by bearoso** [comment]
Hmm. Well I'm still using 16.666 as the automatic frame skip number. Maybe I should fix it. :-)

**Comment by bearoso** [comment]
Wait, shouldn't you be using 60.0988? Nothing is fixed as that 59.94 frequency here. The monitor varies and the SNES is either 60.0988 or 59.984 for interlaced mode.

**Comment by bearoso** [comment]
I put up a hot-fix release for my automatic input rate thing breaking on high refresh rates. I was able to build the Windows binaries.

**Comment by OV2** [comment]
I've updated the mirror. The x64 version seems to be built without ddraw, but I think we can leave it like that for now. Last time I forgot it for both version, so I'll definitely try to get it to automatically build. I think it's only a .lib file that is missing in newer sdks.

**Comment by bearoso** [comment]
Yeah stupid me. I didn't enable the define before compiling it. Inevitably someone's going to complain. I'll do another combined zip like you did last time.

I was looking over the syncing code in Windows, and the auto frameskip stuff is still running with Sound Sync enabled. In the GTK+ port we just disable all the other syncing when sound sync runs (at least, without Dynamic Rate Control enabled). As a user, the first thing I do when I configure is set frameskip to 0, so I didn't notice this before. But I'm realizing that end-users don't seem to tweak the settings as much as I thought they did.

**Comment by OV2** [comment]
Yes, you can combine auto framerate with sound sync. Would help on very low end systems where it can skip drawing frames if it is too slow, but the main use of auto framerate is simply for speed up / slow down to work. That's why it's set to automatic with 0 skips allowed by default instead of fixed with 0, which disables all the frame timer shenanigans. I think I tried that and people were complaining they could no longer change the speed.

**Comment by bearoso** [comment]
Oh, well it interacts badly for me and makes the frame rate all skippy.

**Comment by bearoso** [comment]
@OV2, would you update your snes9x.com post to point to 1.56.1? 

Then, I think we can close this issue.

**Comment by OV2** [comment]
Unfortunately I can no longer edit the post, but I've asked ryan to copy the contents of the new post above the old one.

**Comment by Felipefpl** [comment]
Ipher not updated yet http://snes9x.ipherswipsite.com/

**Comment by Felipefpl** [comment]
Ipher not updated yet http://snes9x.ipherswipsite.com/

Just a simple kinda useless opinion: the purple icon is UGLY, the colorful one even if kinda a clone of the icon of zsnes is way better. Also. for some reason i cant understand the names of the authors of the changes come mixed with the changes in the changelog.

**Comment by OV2** [comment]
We have no control over ipher's site, someone would have to contact him.

The icon colors are the colors of the american controller / SNES. I'm not sure if it's worth it to make the icon configurable.

I'm not sure what you mean about the changelog - every entry has the participants in the second column.

**Comment by bearoso** [comment]
@Felipefpl Here is the non-purple icon we had before:
[snes9x-blue-green-yellow-red-icon.zip](https://github.com/snes9xgit/snes9x/files/2088017/snes9x-blue-green-yellow-red-icon.zip)
@OV2 I think you can include another icon in the exe as a resource, and Windows' icon picker can find it. We could do that.

**Comment by bearoso** [comment]
Of course, that wouldn't affect the application window.

**Comment by ImSpecial** [comment]
For what it's worth, I like the new purple icon, and the colored one looks like the candy [smarties](https://kathonewmedia.files.wordpress.com/2015/01/smarties.jpg) so I think that one is a worse choice.

And the changelog makes sense to me...

**Comment by Felipefpl** [comment]
bearoso - thanks for the icon, much nicer to look at and the 2nd icon is a good idea.

ov2 - i meant this:

- Win32: Cheats are now shown in the correct order in the   (BearOso)
  list view.

Instead of:

- Win32: Cheats are now shown in the correct order in the  list view. (BearOso)

The name of the author comes after the dot in the end. Another good idea is to list the updates under only 1 name for example:

Bearoso's changes:

1 - fixed Sdd1
2- fixed aero acrobat
3 - fixed super-fx

And so on. ;)

**Comment by ImSpecial** [comment]
Why not make a [Pull Request](https://github.com/snes9xgit/snes9x/pulls) about the changelog then? The Emulation world can always use more Readme developers.
